sql,opt: support lookup joins that perform reverse scans

DrewKimball opened the pull request:
#### sql,kvstreamer: enable reverse scans in lookup joins

This commit makes the execution changes necessary to allow lookup joins
to perform lookups as reverse scans rather than forward scans. This mostly
involves making the streamer capable of handling reverse scans, since the
`DistSender` already knows how to handle them. This commit also adds a
`ReverseScans` field to the lookup join spec and plumbs that information
to the fetcher implementations. A follow-up will determine when and how
to set the field.

Informs #128702

Release note: None

#### opt/ordering: add support for reverse scans for lookup joins

This commit adds support to the lookup-join ordering logic for using
reverse scans to provide an ordering opposite the index ordering.
The following commit will plumb support for this through the optimizer.

Informs #128702

Release note: None

#### sql,opt: add support for lookup joins that perform reverse scans

This commit adds support for planning lookup joins that perform reverse
scans for each lookup. This will allow the optimizer to avoid planning
sorts in more cases. This is particularly useful for generic query plans,
which plan lookup joins in place of scans.

Fixes #128702

Release note: None

#### opt: check cluster version and session setting for reverse-scan lookup joins

This commit adds a session setting to toggle planning for lookup joins
that perform reverse scans: `optimizer_plan_lookup_joins_with_reverse_scans`.
A lookup join with reverse scans will only be planned if this setting
is enabled (default on) and the cluster version is at least v25.2.

Informs #128702

Release note: None

DrewKimball committed:
sql,kvstreamer: enable reverse scans in lookup joins

This commit makes the execution changes necessary to allow lookup joins
to perform lookups as reverse scans rather than forward scans. This mostly
involves making the streamer capable of handling reverse scans, since the
`DistSender` already knows how to handle them. This commit also adds a
`ReverseScans` field to the lookup join spec and plumbs that information
to the fetcher implementations. A follow-up will determine when and how
to set the field.

Informs #128702

Release note: None

DrewKimball committed:
opt/ordering: add support for reverse scans for lookup joins

This commit adds support to the lookup-join ordering logic for using
reverse scans to provide an ordering opposite the index ordering.
The following commit will plumb support for this through the optimizer.

Informs #128702

Release note: None

DrewKimball committed:
sql,opt: add support for lookup joins that perform reverse scans

This commit adds support for planning lookup joins that perform reverse
scans for each lookup. This will allow the optimizer to avoid planning
sorts in more cases. This is particularly useful for generic query plans,
which plan lookup joins in place of scans.

Fixes #128702

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142169)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice! It also might good for someone else to review the 2nd commit since it's outside of my area of expertise.

> Reviewed 17 of 17 files at r1, 3 of 3 files at r2, 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-:-OKP9sc90CavQ-ucPScW:b-34pg0m)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/rowcontainer/kvstreamer_result_disk_buffer.go` line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP7FbzAEGY-8erGiNM:-OKP7FbzAEGY-8erGiNN:busfx81) ([raw file](https://github.com/cockroachdb/cockroach/blob/91c57fc2baa916967a480e8f164e056dd210aafe/pkg/sql/rowcontainer/kvstreamer_result_disk_buffer.go#L228)):*
> ```Go
> // 'scanComplete' fields are left unchanged since those aren't serialized.
> //
> // -reverse, if true, indicates that ReverseScanResponse is to be used instead
> ```

nit: add a space: `- reverse`.
___
*[`pkg/kv/kvclient/kvstreamer/streamer.go` line 1644 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP5Wxd3sypGbgGXbQ-:-OKP5Wxd3sypGbgGXbQ0:b-xnesou) ([raw file](https://github.com/cockroachdb/cockroach/blob/91c57fc2baa916967a480e8f164e056dd210aafe/pkg/kv/kvclient/kvstreamer/streamer.go#L1644)):*
> ```Go
> 			if len(GetScanRows(reply)) > 0 {
> 				return fp, errors.AssertionFailedf(
> 					"unexpectedly got a ScanResponse using KEY_VALUES response format",
> ```

nit: maybe adjust the error message with `%T` directive, to avoid hard-coded `ScanResponse`.
___
*[`pkg/kv/kvclient/kvstreamer/util.go` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP4TKS9Q_Q0qTB8Vv2:-OKP4TKS9Q_Q0qTB8Vv3:b6o7bk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/91c57fc2baa916967a480e8f164e056dd210aafe/pkg/kv/kvclient/kvstreamer/util.go#L11)):*
> ```Go
> // GetScanRows returns the Rows field from a ScanResponse or
> // ReverseScanResponse.
> func GetScanRows(response kvpb.Response) (rows []roachpb.KeyValue) {
> ```

nit: only `GetScanBatchResponses` is used outside of `kvstreamer` package. I could see a minor improvement achieved via one of the following:
- keep all methods exported but move the file into `kvpb` to make it more easily findable
- unexport all methods except for `GetScanBatchResponses`, and keep everything in `kvstreamer` given that two usages of the method are in the streamer-related files.
___
*[`pkg/sql/opt/ordering/lookup_join.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP7s6J9nMB8MxjJ626:-OKP7s6J9nMB8MxjJ627:b-y9b0wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f21ddada64e00a01659ae9ba5d1453461c0e0f34/pkg/sql/opt/ordering/lookup_join.go#L29)):*
> ```Go
> ) (canProvide bool, direction ScanDirection) {
> 	if required.Any() {
> 		// LookupJoin an always provide the empty ordering.
> ```

nit: `s/an/can/`.
___
*[`pkg/sql/opt/ordering/lookup_join.go` line 319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP8lA75KR-4s-N9sRq:-OKP8lA75KR-4s-N9sRr:b-w24zz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f21ddada64e00a01659ae9ba5d1453461c0e0f34/pkg/sql/opt/ordering/lookup_join.go#L319)):*
> ```Go
> // * An OrderingChoice that indicates how the prefix needs to be extended in
> // order to imply the required ordering choice.
> // * The direction in which each lookup must scan the index. DefaultDirection
> ```

nit: `s/DefaultDirection/EitherDirection/`?
___
*[`pkg/sql/opt/ordering/lookup_join.go` line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP8zHT9m4oSl-A2L1W:-OKP8zHT9m4oSl-A2L1X:b1r19d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f21ddada64e00a01659ae9ba5d1453461c0e0f34/pkg/sql/opt/ordering/lookup_join.go#L326)):*
> ```Go
> // mutate.
> //
> // - requiredDirection is the direction in which the index must be scanned. Both
> ```

nit: this seems like a leftover?
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2808 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP9NPf5CldKo8nlnOr:-OKP9NPf5CldKo8nlnOs:bmsjztd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d72fa6f0101d361a81b98e9555d4a79c82461fe5/pkg/sql/opt/exec/execbuilder/relational.go#L2808)):*
> ```Go
> 		return execPlan{}, colOrdMap{}, err
> 	}
> 	ok, requiredDirection := ordering.LookupJoinCanProvideOrdering(
> ```

We should only use the newly-added execution support for reverse scans in the lookup joins once the cluster is at 25.2 cluster version.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[`pkg/kv/kvclient/kvstreamer/streamer.go` line 1644 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP5Wxd3sypGbgGXbQ-:-OK_Ws0u9VJBpozu2AT9:b-574kmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/91c57fc2baa916967a480e8f164e056dd210aafe/pkg/kv/kvclient/kvstreamer/streamer.go#L1644)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: maybe adjust the error message with `%T` directive, to avoid hard-coded `ScanResponse`.
</blockquote></details>

Nice, Done.
___
*[`pkg/kv/kvclient/kvstreamer/util.go` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP4TKS9Q_Q0qTB8Vv2:-OK_YnxZ3CZko_bW52re:bprdbwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/91c57fc2baa916967a480e8f164e056dd210aafe/pkg/kv/kvclient/kvstreamer/util.go#L11)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: only `GetScanBatchResponses` is used outside of `kvstreamer` package. I could see a minor improvement achieved via one of the following:
- keep all methods exported but move the file into `kvpb` to make it more easily findable
- unexport all methods except for `GetScanBatchResponses`, and keep everything in `kvstreamer` given that two usages of the method are in the streamer-related files.
</blockquote></details>

I went with the latter, since it seems nice to keep them close to the streamer logic.
___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2808 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP9NPf5CldKo8nlnOr:-OK_ZtLD40Wc1xrihU5d:b-j8m933) ([raw file](https://github.com/cockroachdb/cockroach/blob/d72fa6f0101d361a81b98e9555d4a79c82461fe5/pkg/sql/opt/exec/execbuilder/relational.go#L2808)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We should only use the newly-added execution support for reverse scans in the lookup joins once the cluster is at 25.2 cluster version.
</blockquote></details>

Huh, does this mean that https://github.com/cockroachdb/cockroach/pull/84689 was broken?
___
*[`pkg/sql/opt/ordering/lookup_join.go` line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP7s6J9nMB8MxjJ626:-OK_XGKh2Kz_It6tTbkH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f21ddada64e00a01659ae9ba5d1453461c0e0f34/pkg/sql/opt/ordering/lookup_join.go#L29)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/an/can/`.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/ordering/lookup_join.go` line 319 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP8lA75KR-4s-N9sRq:-OK_XMTcD_m8go8flxfl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f21ddada64e00a01659ae9ba5d1453461c0e0f34/pkg/sql/opt/ordering/lookup_join.go#L319)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/DefaultDirection/EitherDirection/`?
</blockquote></details>

Done.
___
*[`pkg/sql/opt/ordering/lookup_join.go` line 326 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP8zHT9m4oSl-A2L1W:-OK_Xc-_7doL37LKJ5ni:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f21ddada64e00a01659ae9ba5d1453461c0e0f34/pkg/sql/opt/ordering/lookup_join.go#L326)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this seems like a leftover?
</blockquote></details>

Done.
___
*[`pkg/sql/rowcontainer/kvstreamer_result_disk_buffer.go` line 228 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP7FbzAEGY-8erGiNM:-OK_WVQS2lEI3rjnk_ZH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/91c57fc2baa916967a480e8f164e056dd210aafe/pkg/sql/rowcontainer/kvstreamer_result_disk_buffer.go#L228)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: add a space: `- reverse`.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the mixed version state handling.

> Reviewed 18 of 18 files at r4, 3 of 3 files at r5, 12 of 12 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-:-OKbe0OA8s8-_q0vC0fw:b-v9q9h9)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2808 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP9NPf5CldKo8nlnOr:-OKbdXGK7nqiQE1-Oltf:b-l7toxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d72fa6f0101d361a81b98e9555d4a79c82461fe5/pkg/sql/opt/exec/execbuilder/relational.go#L2808)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Huh, does this mean that https://github.com/cockroachdb/cockroach/pull/84689 was broken?
</blockquote></details>

I think in that case we got lucky (or implicitly relied on it) because we had at least one distsql version bump during 22.2 cycle (for example, 3dd1d4f61496912f0a094cb6bfe46863d9507e5c) which effectively disabled distsql in the mixed version state. We refactored the distsql version mechanism, so now we need to consider the cluster version when deciding whether to use new functionality like this one.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Oh, and perhaps it might be nice to add a session variable to disable this feature in the optimizer, just in case.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-:-OKbeG740E-8UhrGrEDB:bi4ertc)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)



<!-- Sent from Reviewable.io -->

DrewKimball committed:
opt: check cluster version and session setting for reverse-scan lookup joins

This commit adds a session setting to toggle planning for lookup joins
that perform reverse scans: `optimizer_plan_lookup_joins_with_reverse_scans`.
A lookup join with reverse scans will only be planned if this setting
is enabled (default on) and the cluster version is at least v25.2.

Informs #128702

Release note: None

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/opt/exec/execbuilder/relational.go` line 2808 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKP9NPf5CldKo8nlnOr:-OKbrJN68OhtwhYwIiHR:b-o8yjwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d72fa6f0101d361a81b98e9555d4a79c82461fe5/pkg/sql/opt/exec/execbuilder/relational.go#L2808)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think in that case we got lucky (or implicitly relied on it) because we had at least one distsql version bump during 22.2 cycle (for example, 3dd1d4f61496912f0a094cb6bfe46863d9507e5c) which effectively disabled distsql in the mixed version state. We refactored the distsql version mechanism, so now we need to consider the cluster version when deciding whether to use new functionality like this one.
</blockquote></details>

I see, thanks for the explanation. Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
opt changes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r12, 8 of 11 files at r13, 17 of 18 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-:-OKdKmVqEMi2OILX6EAC:b-yql68r)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/opt/ordering/lookup_join.go` line 97 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKdIzkP7rToqfrMjw8u:-OKdIzkP7rToqfrMjw8v:br2virk) ([raw file](https://github.com/cockroachdb/cockroach/blob/675732a5af72b6263ec4cacb602138376d53797b/pkg/sql/opt/ordering/lookup_join.go#L97)):*
> ```Go
> 		// Make sure the cluster is at least at v25.2 before planning a lookup
> 		// join with reverse scans.
> 		if !evalCtx.Settings.Version.IsActive(ctx, clusterversion.V25_2) {
> ```

I thought we usually added new cluster versions specific to each feature? Like `V25_2_ReverseScansInLookupJoins` or something (not sure if it matters...)
___
*[`pkg/sql/opt/xform/testdata/physprops/ordering` line 2741 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKdIRthEbODLpmLRso-:-OKdIRti28rai90-Gj4B:b-7sxyul) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b4abcb72370caddf98d2139d92b5487b6eaf515/pkg/sql/opt/xform/testdata/physprops/ordering#L2741)):*
> ```
> SELECT * FROM xyz INNER LOOKUP JOIN abc ON x = a ORDER BY x, y, z, a DESC, b DESC, c DESC
> ----
> inner-join (lookup abc)
> ```

should the output show that this is a reverse scan?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[`pkg/sql/opt/ordering/lookup_join.go` line 97 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKdIzkP7rToqfrMjw8u:-OKeVXdJ8qb99sBeLrGj:b-r39j7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/675732a5af72b6263ec4cacb602138376d53797b/pkg/sql/opt/ordering/lookup_join.go#L97)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I thought we usually added new cluster versions specific to each feature? Like `V25_2_ReverseScansInLookupJoins` or something (not sure if it matters...)
</blockquote></details>

You only need to define a new version gate for things like migrations which need to proceed in steps. Something like this just needs to know that the cluster has finished upgrading, so the `V25_2` placeholder is the right one to use. ([docs here](https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/2775876082/Guidelines+for+writing+cluster+version-specific+code+and+using+version+gates))
___
*[`pkg/sql/opt/xform/testdata/physprops/ordering` line 2741 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKdIRthEbODLpmLRso-:-OKekJjh9TJ-UVAfRUSj:br1snl2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b4abcb72370caddf98d2139d92b5487b6eaf515/pkg/sql/opt/xform/testdata/physprops/ordering#L2741)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should the output show that this is a reverse scan?
</blockquote></details>

Good point - it required introducing a helper function for expr formatting, but it's simplified by the fact that we can rely on the required ordering to be satisfiable by the lookup join.

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 12 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-:-OKfxn5E0GWN0ssChFwc:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @yuzefovich)

___
*[`pkg/sql/opt/ordering/lookup_join.go` line 97 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/142169#-OKdIzkP7rToqfrMjw8u:-OKfwfdj6-obO1gcavvE:b-ytwisq) ([raw file](https://github.com/cockroachdb/cockroach/blob/675732a5af72b6263ec4cacb602138376d53797b/pkg/sql/opt/ordering/lookup_join.go#L97)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

You only need to define a new version gate for things like migrations which need to proceed in steps. Something like this just needs to know that the cluster has finished upgrading, so the `V25_2` placeholder is the right one to use. ([docs here](https://cockroachlabs.atlassian.net/wiki/spaces/ENG/pages/2775876082/Guidelines+for+writing+cluster+version-specific+code+and+using+version+gates))
</blockquote></details>

Nice, thanks for the pointer


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

bors r=yuzefovich,rytaft
